pkg/controller/inferenceservice/controller_test.go

```diff
@@ -1136,3 +1136,163 @@ func TestInferenceServiceWithTransformer(t *testing.T) {
+			Name:      constants.InferenceServiceConfigMapName,
+			Namespace: constants.KFServingNamespace,
+		},
+		Data: configs,
+	}
+	g.Expect(c.Create(context.TODO(), configMap)).NotTo(gomega.HaveOccurred())
+	defer c.Delete(context.TODO(), configMap)
+
+	// Create the InferenceService object and expect the Reconcile and knative service to be created
+	instance := transformer.DeepCopy()
+	g.Expect(c.Create(context.TODO(), instance)).NotTo(gomega.HaveOccurred())
+	defer c.Delete(context.TODO(), instance)
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
+
+	defaultPredictorService := &knservingv1alpha1.Service{}
+	g.Eventually(func() error { return c.Get(context.TODO(), defaultPredictor, defaultPredictorService) }, timeout).
+		Should(gomega.Succeed())
+
+	canaryPredictorService := &knservingv1alpha1.Service{}
+	g.Eventually(func() error { return c.Get(context.TODO(), canaryPredictor, canaryPredictorService) }, timeout).
+		Should(gomega.Succeed())
+
+	defaultTransformerService := &knservingv1alpha1.Service{}
+	g.Eventually(func() error { return c.Get(context.TODO(), defaultTransformer, defaultTransformerService) }, timeout).
+		Should(gomega.Succeed())
+
+	canaryTransformerService := &knservingv1alpha1.Service{}
+	g.Eventually(func() error { return c.Get(context.TODO(), canaryTransformer, canaryTransformerService) }, timeout).
+		Should(gomega.Succeed())
+
+	// Update instance to remove transformer endpoint
+	// transformer services should be removed during reconcile
+	updateInstance := &kfserving.InferenceService{}
+	g.Eventually(func() error { return c.Get(context.TODO(), serviceKey, updateInstance) }, timeout).
+		Should(gomega.Succeed())
+
+	updateInstance.Spec.Canary.Transformer = nil
+	updateInstance.Spec.Default.Transformer = nil
+	g.Expect(c.Update(context.TODO(), updateInstance)).NotTo(gomega.HaveOccurred())
+
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
+
+	defaultTransformerServiceShouldBeDeleted := &knservingv1alpha1.Service{}
+	g.Eventually(func() bool {
+		err := c.Get(context.TODO(), defaultTransformer, defaultTransformerServiceShouldBeDeleted)
+		return errors.IsNotFound(err)
+	}, timeout).Should(gomega.BeTrue())
+
+	canaryTransformerServiceShouldBeDeleted := &knservingv1alpha1.Service{}
+	g.Eventually(func() bool {
+		err := c.Get(context.TODO(), canaryTransformer, canaryTransformerServiceShouldBeDeleted)
+		return errors.IsNotFound(err)
+	}, timeout).Should(gomega.BeTrue())
+	defaultPredictorServiceShouldExist := &knservingv1alpha1.Service{}
+	g.Eventually(func() error { return c.Get(context.TODO(), defaultPredictor, defaultPredictorServiceShouldExist) }, timeout).
+		Should(gomega.Succeed())
+
+	canaryPredictorServiceShouldExist := &knservingv1alpha1.Service{}
+	g.Eventually(func() error { return c.Get(context.TODO(), canaryPredictor, canaryPredictorServiceShouldExist) }, timeout).
```

Comment:
Nice test! Can you help add a check if transformer ready condition is removed also ?

Reply:
sure, done